cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go

```diff
@@ -0,0 +1,132 @@
+		Long:  cmdutil.MacroCommandLongDescription,
+		Phases: []workflow.Phase{
+			newCertsSubphase(),
+			newKubeconfigSubphase(),
+			newManifestsSubphase(),
+		},
+	}
+}
+
+func newCertsSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "certs",
+		Short: "Generates the certificates for all the new control plane components",
+		Run:   runCertsPhaseLocal,
+	}
+}
+
+func newKubeconfigSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "kubeconfig",
+		Short: "Generates the kubeconfig for all the new control plane components",
+		Run:   runKubeconfigPhaseLocal,
+	}
+}
+
+func newManifestsSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "manifests",
+		Short: "Generates the manifests for all the new control plane components",
+		Run:   runManifestsPhaseLocal,
+	}
+}
+
+func runCertsPhaseLocal(c workflow.RunData) error {
+	data, ok := c.(controlPlanePrepareData)
+	if !ok {
+		return errors.New("certs phase invoked with an invalid data struct")
+	}
+
+	// Generate missing certificates (if any)
+	return certsphase.CreatePKIAssets(data.Cfg())
+}
+
+func runKubeconfigPhaseLocal(c workflow.RunData) error {
+	data, ok := c.(controlPlanePrepareData)
+	if !ok {
+		return errors.New("kubeconfig phase invoked with an invalid data struct")
+	}
+
+	// Generate kubeconfig files for controller manager, scheduler and for the admin/kubeadm itself
+	// NB. The kubeconfig file for kubelet will be generated by the TLS bootstrap process in
+	// following steps of the join --experimental-control plane workflow
+	if err := kubeconfigphase.CreateJoinControlPlaneKubeConfigFiles(kubeadmconstants.KubernetesDir, data.Cfg()); err != nil {
+		return errors.Wrap(err, "error generating kubeconfig files")
+	}
+
+	return nil
+}
+
+func runManifestsPhaseLocal(c workflow.RunData) error {
+	data, ok := c.(controlPlanePrepareData)
+	if !ok {
+		return errors.New("manifests phase invoked with an invalid data struct")
+	}
+	cfg := data.Cfg()
+
+	// Creates static pod manifests file for the control plane components to be deployed on this node
+	// Static pods will be created and managed by the kubelet as soon as it starts
+	if err := controlplanephase.CreateInitStaticPodManifestFiles(kubeadmconstants.GetStaticPodDirectory(), cfg); err != nil {
+		return errors.Wrap(err, "error creating static pod manifest files for the control plane components")
+	}
+
+	if cfg.Etcd.External == nil {
+		// Checks that the etcd cluster is healthy
+		// NB. this check cannot be implemented before because it requires the admin.conf and all the certificates
+		//     for connecting to etcd already in place
+		kubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName)
+
+		client, err := kubeconfigutil.ClientSetFromFile(kubeConfigFile)
+		if err != nil {
+			return errors.Wrap(err, "couldn't create Kubernetes client")
+		}
+
+		if err := etcdphase.CheckLocalEtcdClusterStatus(client, cfg); err != nil {
+			return err
```

Comment:
we might have coherence/guideline about errors. `should we wrap all errors`  or return them?

Reply:
e.g should we add `[join/phaseName]` in the error string for being more precise?

Reply:
I guess that `[phaseName]` is a must have